src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/ApplicationIdRetriever.java

```diff
@@ -0,0 +1,114 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.bigquery.dwhassessment.hooks.logger;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Optional;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.llap.registry.impl.LlapRegistryService;
+import org.apache.hadoop.hive.ql.MapRedStats;
+import org.apache.hadoop.hive.ql.session.SessionState;
+import org.apache.hadoop.mapred.JobID;
+import org.apache.hadoop.mapreduce.TypeConverter;
+import org.apache.hadoop.yarn.api.records.ApplicationId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Helper for retrieving YARN Application ID depending on the execution mode. */
+public class ApplicationIdRetriever {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ApplicationIdRetriever.class);
+
+  public static Optional<ApplicationId> determineApplicationId(
+      HiveConf conf, ExecutionMode executionMode) {
+    switch (executionMode) {
+      case MR:
+        return determineMapReduceApplicationId();
+      case TEZ:
+        return determineTezApplicationId();
+      case LLAP:
+        return determineLlapApplicationId(conf, executionMode);
+      default:
+        return Optional.empty();
+    }
+  }
+
+  /**
+   * Retrieves Application ID for the Tez application. Application can be reused, but one
+   * application always have only one queue – if queue changes in the session, new application is
+   * created.
+   */
+  private static Optional<ApplicationId> determineTezApplicationId() {
+    try {
+      return Optional.of(
+          SessionState.get().getTezSession().getSession().getAppMasterApplicationId());
+    } catch (NullPointerException e) {
+      LOG.info("Failed to retrieve Application ID from Tez session");
+      return Optional.empty();
+    }
+  }
+
+  /**
+   * Retrieves Application ID from the first MapReduce job as multiple MapReduce jobs created for a
+   * single query are submitted to the same queue.
+   */
+  private static Optional<ApplicationId> determineMapReduceApplicationId() {
+    return SessionState.get().getMapRedStats().values().stream()
+        .map(MapRedStats::getJobId)
+        .map(
+            jobId -> {
+              try {
+                return TypeConverter.toYarn(JobID.forName(jobId)).getAppId();
+              } catch (IllegalArgumentException e) {
+                LOG.warn(
+                    "Failed to convert MapReduce Job ID '{}' to YARN application id, this is"
+                        + " unexpected.",
+                    jobId);
+                return null;
+              }
+            })
+        .filter(Objects::nonNull)
```

Comment:
Instead of filtering null, I would suggest replacing the above `map` call with `flatMap` and return `Stream.of(...)` and `Stream.empty()` accordingly.

Reply:
Done